keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -638,24 +640,153 @@ always dispatched immediately.  Following is how the other requests
-priority level:
+priority levels, and higher ones can do a limited amount of borrowing
+from lower ones.
 
-```
-ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+Two fields of `LimitedPriorityLevelConfiguration`, introduced in the
+midst of the `v1beta2` lifetime, configure the borrowing.  The following
+display shows the two new fields along with the updated description for
+the `AssuredConcurrencyShares` field.
+
+```go
+type LimitedPriorityLevelConfiguration struct {
+  ...
+  // `assuredConcurrencyShares` (ACS) contributes to the computation of the
+  // NominalConcurrencyLimit (NCL) of this level.
+  // This is the number of execution seats available at this priority level.
+  // This is used both for requests dispatched from
+  // this priority level as well as requests dispatched from higher priority
+  // levels borrowing seats from this level.  This does not limit dispatching from
+  // this priority level that borrows seats from lower priority levels (those lower
+  // levels do that).  The server's concurrency limit (SCL) is divided among the
+  // Limited priority levels in proportion to their ACS values:
+  //
+  // NCL(i)  = ceil( SCL * ACS(i) / sum_acs )
+  // sum_acs = sum[limited priority level k] ACS(k)
+  //
+  // Bigger numbers mean a larger nominal concurrency limit, at the expense
+  // of every other Limited priority level.
+  // This field has a default value of 30.
+  // +optional
+  AssuredConcurrencyShares int32
+
+  // `lendableConcurrencyShares` (LCS) contributes to the computation of the
+  // LendableConcurrencyLimit (LCL) for this level.  This is the number of
+  // execution seats of this level that can be borrowed by higher priority
+  // Limited levels.
+  // This may not be negative, and may not be greater than
+  // `assuredConcurrencyShares`.
+  //
+  // LCL(i) = ceil( SCL * LCS(i) / sum_acs )
+  //
+  // This field has a default value of zero.
+  // +optional
+  LendableConcurrencyShares int32
+
+  // `priority` determines where this priority level appears in the total order
+  // of Limited priority levels used to configure borrowing between those levels.
+  // A numerically higher value means a logically lower priority.
+  // Do not create ties; they will be broken arbitrarily.
+  // `priority` SHOULD be a positive number and MUST NOT be greater than 10000.
+  // If it is zero then, for the sake of a smooth transition from the time
+  // before this field existed, this level will be treated as if its `priority`
+  // is the average of the `matchingPrecedence` of the FlowSchema objects
+  // that reference this level.
+  // +optional
+  Priority int32
+}
 ```
 
-where SCL is the apiserver's concurrency limit and ACS(l) is the
-AssuredConcurrencyShares for priority level l.
+This is a somewhat tortured meaning for "assured", but it is the
+meaning we need for introduction of the new field to the existing type
+while having a smooth transition in behavior.  In the next version we
```

Comment:
You are thinking, if they were additive, and we make lendable non-zero by default for some levels, and don't adjust the global concurrency setting, then suddenly a share is smaller than it was and levels will get a different effective amount of concurrency than before?

I think that is not actually a problem: user-created PLs will all have lendable set to zero, and thus no change. And if we want to make the default PLs have non-zero lendable values by default, we can simultaneously reduce the assured values, so that the net change in shares is zero.

(I wouldn't be mentioning this were it not for your good point that we will have made "assured" mean something unnatural otherwise)

Reply:
We are squeezed by multiple considerations, when adding a field to an existing type.  Remember that we need to be able to do incremental rollout of new servers into a cluster with old servers.  We have decided that servers with different opinions will just repeatedly overwrite the objects until the rollout has finished.  Remember also that APF's config-consuming controller (the thing that digests config objects into the data structure used for queuing and dispatching) is independent from APF's config-producing controller.  So what happens in an old server when it digests a revised object, not seeing or understanding the new fields?  It simply processes the new values of the old fields.  We do not want that to produce a big behavior change in the old server.  So that tells me that we do not want this PR to propose changing the values of `AssuredConcurrencyShares` (at least, not their relative magnitudes, but I do not see that as helping any, since the other shares are also relative).

Similarly, we also have to think about what happens when a new server digests an old-content object, possibly intermixed with new-content objects.  Remember that an informer guarantees a client no ordering between objects.  That is what led to the proposal that the behavior when `Priority=0` is as if `Priority` were the average of the related `MatchingPrecedence` values.

Reply:
Recent revisions reflect agreement on roadmap.